addons/addon-base-raas/packages/base-raas-services/lib/data-egress/data-egress-service.js

```diff
@@ -0,0 +1,236 @@
+
+    return egressStore;
+  }
+
+  async audit(requestContext, auditEvent) {
+    const auditWriterService = await this.service('auditWriterService');
+    // Calling "writeAndForget" instead of "write" to allow main call to continue without waiting for audit logging
+    // and not fail main call if audit writing fails for some reason
+    // If the main call also needs to fail in case writing to any audit destination fails then switch to "write" method as follows
+    // return auditWriterService.write(requestContext, auditEvent);
+    return auditWriterService.writeAndForget(requestContext, auditEvent);
+  }
+
+  // @private
+  async getKmsKeyIdArn() {
+    // Get the kms key id
+    const kmsAliasArn = this.settings.get(settingKeys.egressStoreKmsKeyAliasArn);
+
+    // Get KMS Key ARN from KMS Alias ARN
+    // The "Decrypt","DescribeKey","GenerateDataKey" etc require KMS KEY ARN and not ALIAS ARN
+    const kmsClient = await this.getKMS();
+    const data = await kmsClient
+      .describeKey({
+        KeyId: kmsAliasArn,
+      })
+      .promise();
+    return data.KeyMetadata.Arn;
+  }
+
+  async getKMS() {
+    const aws = await this.getAWS();
+    return new aws.sdk.KMS();
+  }
+
+  async getAWS() {
+    const aws = await this.service('aws');
+    return aws;
+  }
+
+  async getS3() {
+    const aws = await this.getAWS();
+    return new aws.sdk.S3();
+  }
+
+  async getS3BucketAndPolicy() {
+    const s3BucketName = this.settings.get(settingKeys.egressStoreBucketName);
+    const s3Client = await this.getS3();
+    const s3Policy = JSON.parse((await s3Client.getBucketPolicy({ Bucket: s3BucketName }).promise()).Policy);
+    if (!s3Policy.Statement) {
+      s3Policy.Statement = [];
+    }
+    return { s3BucketName, s3Policy };
+  }
+
+  async addEgressStoreBucketPolicy(requestContext, egressStore, memberAccountId) {
+    const { s3BucketName, s3Policy } = await this.getS3BucketAndPolicy();
+
+    const statementParamFunctions = [];
+    if (egressStore.envPermission.read) {
+      statementParamFunctions.push(getStatementParamsFn);
+    }
+    if (egressStore.envPermission.write) {
+      statementParamFunctions.push(putStatementParamsFn);
+    }
+    if (egressStore.envPermission.read || egressStore.envPermission.write) {
+      statementParamFunctions.push(listStatementParamsFn);
+    }
+    const revisedStatements = await this.getRevisedS3Statements(
+      s3Policy,
+      egressStore,
+      s3BucketName,
+      statementParamFunctions,
+      oldStatement => addAccountToStatement(oldStatement, memberAccountId),
+    );
+
+    const s3Client = await this.getS3();
+
+    await updateS3BucketPolicy(s3Client, s3BucketName, s3Policy, revisedStatements);
+
+    // Write audit event
+    await this.audit(requestContext, { action: 'add-egress-store-bucket-policy', body: s3Policy });
+  }
+
+  // @private
+  async getRevisedS3Statements(s3Policy, egressStore, bucket, statementParamFunctions, updateStatementFn) {
```

Comment:
Seems like a duplicate method in environment-resource-service.js. Can you move it to utils.

Reply:
Resolved in commit: be8afee164ee1b0b18c86c817753979989a273dd